README.MD

# PTTYDev - Pseudo TTY Device

PTTYDev provides a common interface and enhanced timeout and automatic reconnection handling.
Using a pttydev enables your code to read / write a serial device or websocket.

## Platform

Tested on ESP32, pyserial and MicroPython WebRepl

## Related project

[mpycntrl](https://github.com/kr-g/mpycntrl)


## Development status

Alpha state,

in case of unexpected behaviour change timeouts for your board


### Limitation

deprecated hint: 
there is a known bug with [WebREPL](https://github.com/micropython/micropython/issues/2497)


## Installation

pttydev can be installed with pip from here [pttydev](https://pypi.org/project/pttydev/)


## Sample Code

some sample code is under the pttydev directory on project homepage on github

[sample.py](https://github.com/kr-g/pttydev/blob/master/pttydev/sample.py)





CHANGELOG.MD

# Changelog

# v0.0.4

added 0.5 sec delay between sending chunks of 255 bytes via WEBRepl

tested on ESP8266, and ESP32 with PSRam

https://github.com/micropython/micropython/issues/2497





pttydev/dev_ws.py

# https://github.com/websocket-client/websocket-client
#import websocket

from websocket import *
import socket
import time

    
class _WebSocket_Dev():
    
    """context wrapper for websocket device"""
    
    def __init__(self,ws,write_delay):
        self.ws = ws
        self.write_delay = write_delay
        
    def __enter__(self):
        return self
    
    def __exit__(self, exception_type, exception_value, traceback):
        import traceback
        print( exception_type, exception_value, traceback )
        self.ws.close()
        
    def read(self,size=1):
        r = None
        try:
            r = self.ws.recv()
        except WebSocketTimeoutException:
            pass
            #print("dev timeout")
        if r is None:
            return r
        try:
            return r.encode()
        except:
            return r
        
    def write(self,buf):
        #print("dev write", buf )
        
        wbtotal = 0
        
        for i in range(0, len(buf), 255):
            bslice = buf[i:min(i + 255, len(buf))]
            wb = self.ws.send(bslice)
            wbtotal += wb
            print("dev write slice", wb, len(bslice) )
            time.sleep(self.write_delay)
        
        print("dev write total", wbtotal, len(buf) )
        return wbtotal
    
    def flush(self):
        import os
        fd = self.ws.sock.fileno()
        #print("dev flush fd", self.ws.sock, fd)
              
   


def pttywsopen(url,password,timeout=3,write_delay=0.5):
    
    # a open function is required to reconnect properly the device in case of error
    #websocket.enableTrace(False)
   
    ws = create_connection(url,timeout=timeout,
                           sockopt=((socket.IPPROTO_TCP, socket.TCP_NODELAY, 0),)
                                   )        
    r =  ws.recv()
    print("dev got", r )
    if r.find("Password:")>=0:
        ws.send(password+"\n")
    r = ws.recv()
    print("dev got", r )
    if r.find("WebREPL connected")>=0:
        ptty = _WebSocket_Dev(ws,write_delay)
        return ptty
    
    raise Exception( "wrong password, additional information", r)



pttydev/pttydev.py
from collections.abc import Sized
from threading import Thread
import time

from atomic import *


def peekone(buf,matching):
    pos = -1
    for m in matching:
        match = buf.find( m )
        if pos==-1 and match>=0:
            pos=len(buf)
        pos = min( pos, match )
    return pos

def peekone_r(buf,matching):
    pos = -1
    for m in matching:
        match = buf.rfind( m )
        pos = max( pos, match )
    return pos


class PseudoTTYException(Exception):
    def __init__(self,*args ):
        super().__init__(*args)
        
class PseudoTTYBufferException(Exception):
    pass

class PseudoTTYOpenException(Exception):
    pass

class PseudoTTYTimeoutException(Exception):
    pass


def _the_thread_reader( ptty ):
    
    ptty._print_d("thread startup")
    ptty.thrd_stat = 0 # thread starting
    
    while True:
        try:
            ptty._print_d("thread connecting")
            with ptty.pttyopen() as ptty_dev:
                
                ptty._print_d("thread connected", type(ptty_dev))
                ptty.dev = ptty_dev
                
                ptty.thrd_stat = 1 # device is open
                buf = bytearray()
                
                while True:
                    if ptty.status < 0:
                        ptty._print_d("thread shutdown")
                        ptty.thrd_stat = None
                        return
                    
                    stoptime = time.time()+0.1
                    readmax=1
                    while readmax>0 and time.time() < stoptime:
                        readmax-=1
                        cb = ptty_dev.read(ptty.block_size)
                        if cb is None:
                            #raise PseudoTTYTimeoutException("device timeout")
                            break # timeout
                        if len(cb)>0:
                            ptty._print_d("thread received serial", cb, debug=ptty._thrd_debug)
                            buf.extend(cb)
                        else:
                            break
                    
                    try:
                        ptty.addtoibuf( buf )
                        buf.clear()
                    except AtomicOwnerException:
                        ptty._print_d("skip serial read",debug=ptty._thrd_debug)
                                            
        except Exception as ex:
            ptty.dev = None
            ptty.dev_reconnect_count += 1
            ptty._print_d("reconnect", ex )
            import traceback
            traceback.print_exception(None, ex, ex.__traceback__)
            ptty.thrd_stat = 2 # device is reconnecting
            ptty._print_t("thread sleep between reconnect", trace=ptty._thrd_trace_on)
            time.sleep(ptty.reconnect_delay)
     
          
class _Open_Context():

    """context wrapper for ptty device"""

    def __init__(self,ptty):
        self.ptty = ptty

    def __enter__(self):
        self.ptty._open()
        return self.ptty

    def __exit__(self, exception_type, exception_value, traceback):
        self.ptty._close()
        

class PseudoTTY(Atomic,Sized):
    
    DEFAULT_READLINE_DELIMITER = [0x00,0x01,0x02,0x03,0x04,0x05,ord("\r"),ord("\n")]

    def __init__(self,pttyopen,
                 thrd_reader=_the_thread_reader,
                 timeout=0.1,
                 block_size=512,
                 reconnect_delay=1,
                 trace_on=False,debug=False,
                 thrd_trace_on=False,thrd_debug=False
                 ):
        
        super().__init__(trace_on=not True)
        
        self._ibuf = bytearray()
        
        self.timeout = timeout
        self.block_size = block_size
        self.pttyopen = pttyopen
        self.dev = None
        self.dev_reconnect_count = 0
        self.dev_reconnect_time = None
        
        self.status = 0
        self.reconnect_delay = reconnect_delay
        
        self.thrd_reader = thrd_reader #if thrd_reader != None else _the_thread_reader
        self.thrd = None
        self.thrd_stat = None
        self.thrd_timeout = self.timeout / 100
        
        self._main_trace_on = trace_on
        self._main_debug = debug
        self._thrd_trace_on = thrd_trace_on
        self._thrd_debug = thrd_debug

    #
    
    def _print_t( self, *args, trace=None ):
        if trace is None and self._main_trace_on:
            trace = True
        if trace:
             print("(trace)", *args )

    def _print_d( self, *args, debug=None ):
        if debug is None and self._main_debug:
            debug = True
        if debug:
             print("(debug)", *args )

    #
    
    def open(self):
        return _Open_Context(self)
    
    def _open(self):
        if self.thrd:
            raise( PseudoTTYOpenException("the device is already open") )
        self.thrd = Thread( None, _the_thread_reader, args=[self] )
        self.thrd_stat = None
        self.thrd.start()
    
    def _close(self):
        """close ptty thread down by setting status to -1"""
        if self.thrd is None:
            raise( PseudoTTYOpenException("the device is not open") )
        self.status = -1
        self.thrd.join()
        self.thrd = None
        
    def close(self):
        return self._close()
        
    # basic buffer functions
    
    @Atomic.LockFunc
    def reset_input_buffer(self):
        """clear the intenal buffer"""
        self._print_d("clr buffer",self._ibuf)
        self._ibuf.clear()
    
    @Atomic.LockFunc
    def addtoibuf(self,data):
        """add data to internal buffer"""
        self._ibuf.extend(data)

    @Atomic.LockFunc
    def getbufcopy(self):
        """get a copy of the internal buffer"""
        return bytearray(self._ibuf)
    
    @Atomic.LockFunc
    def available(self):
        """get amount of available chars to read from internal buffer"""
        return len(self._ibuf)

    def len(self):
        return self.available()
    
    def __len__(self):
        return self.len()

    # stream functions
    
    @Atomic.LockFunc
    def getch(self):
        """get one char from the stream"""
        try:
            return self._ibuf.pop(0)
        except IndexError:
            return None
        
    @Atomic.LockFunc
    def ungetch(self,ch):
        """unget/pushback one char to the stream"""
        self._ibuf.insert(0,ch)
    
    #

    def peek(self,s):
        """get the position of s within the internal buffer, or -1 if not found"""
        return self.getbufcopy().find(s)

    def peekone(self,matching):
        """find the first occurance of the matching set within the internal buffer, or -1 if not found"""
        return peekone(self._ibuf,matching)

    def peekone_r(self,matching):
        """find the first occurance of the matching set with the internal buffer (searching from the right), or -1 if not found"""
        return peekone_r(self._ibuf,matching)

    #
    
    def waitready(self,timeout=5):
        stop = time.time()+timeout 
        while time.time() < stop:
            if self.dev != None:
                return
            time.sleep(0.1)
        raise PseudoTTYOpenException("device not open")        
    
    #
    
    def write(self,buf):
        if self.dev is None:
            raise PseudoTTYOpenException("device not open")

        self._print_d("ptty write", buf )
        # we assume that the device sends the complete buffer
        wb = self.dev.write(buf)
        # wb might be different from len(buf)
        # eg. websocket will add frame to payload -> wb > len(buf)
        return wb

    def flush(self):
        if self.dev is None:
            raise PseudoTTYOpenException("device not open")
        self.dev.flush()

    @Atomic.LockFunc
    def _readibuf(self,size=1):
        """get chars from the internal buffer"""
        if len(self._ibuf) <= 0:
            r = self._ibuf
            self._ibuf.clear()
            return r
        
        r = self._ibuf[:size]
        self._ibuf = self._ibuf[size:]
        return r

    #             

    MAX_READ_DELAY=0.01

    def read(self,size=1,timeout=0):
        """read a defined length buffer from the internal buffer"""
        delta = max( self.timeout, timeout )
        stop = time.time() + delta
        delay = max(delta / 10, self.MAX_READ_DELAY)
        r = None
        while True:
            try:
                if size<=0:
                    r = self._readibuf(-1)
                elif len(self._ibuf)>=size:
                    r = self._readibuf(size)
                    self._print_d( "read", r )
                    return r 
            except AtomicOwnerException:
                self._print_t("skip read this loop, buffer locked by thread reader")
             
            if time.time() > stop :
                break
            
            self._print_t("delay read for", delay)
            time.sleep(delay)
            
        return r
            
    def readline(self,timeout=0,delimiter=DEFAULT_READLINE_DELIMITER,combine_crlf=True):
        """read one line from the internal buffer"""
        delta = max( self.timeout, timeout )
        stop = time.time() + delta
        delay = max(delta / 10, self.MAX_READ_DELAY)
        r = None
        while True:
            #print( self._ibuf )
            try:
                pos = self.peekone( delimiter )
                if pos >= 0:
                    r = self._readibuf(pos+1)
                    if r != None and len(r)>0:
                        if combine_crlf:
                            if r[-1:][0] == ord("\r"):
                                ch = self.getch()
                                if ch == ord("\n"):
                                    # add look-a-head
                                    r.append(ch)
                                else:
                                    # push back look-a-head
                                    self.ungetch(ch)                        
                    break
            except AtomicOwnerException:
                self._print_t("skip readline this loop, buffer locked by thread reader")                
            if time.time() < stop:
                self._print_t("delay readline for", delay)
                time.sleep(delay)
            else:
                break
                #raise PseudoTTYTimeoutException()
        if r!=None and len(r)>0:
            self._print_d( "readline", r )
        return r     
 
    def readlines(self,timeout=0,delimiter=DEFAULT_READLINE_DELIMITER,combine_crlf=True):
        delta = max( self.timeout, timeout )
        stop = time.time() + delta
        delay = max(delta / 10, self.MAX_READ_DELAY)
        buf = []
        while True:
            b = self.readline(timeout=timeout,delimiter=delimiter,combine_crlf=combine_crlf)
            if b is None:
                break
            buf.append(b)
            if time.time() < stop:
                self._print_t("delay readlines for", delay)
                time.sleep(delay)
            else:
                break
        return buf

pttydev/dev_wsapp.py

# https://github.com/websocket-client/websocket-client
#import websocket

import websocket

from threading import Thread
    
import time
from collections import deque

    
class _WebSocketApp_Dev():
    
    """context wrapper for websocket app device"""
    
    def __init__(self,passwd,timeout=3,debug=False,trace=False, debug_msg=False):
        self.debug = debug
        self.debug_msg = debug_msg
        self.trace = trace
        
        self.passwd = passwd
        self.timeout = timeout
        self.ws = None
        self.buf = deque()
        self.readbuf = bytearray()
        
        self.connected = None
        
    def print_d(self,*args):
        if self.debug:
            print("(debug)",*args)
            
    def print_t(self,*args):
        if self.trace:
            print("(trace)",*args)
    
    
    def __enter__(self):
        self.print_t("enter")
        return self
    
    def __exit__(self, exception_type, exception_value, traceback):
        self.print_t("exit")
        self.close()
       
       
    def close(self):
        self.print_t("close")
        self.ws.keep_running = False

    def clearbuf(self):
        self.readbuf.clear()
        self.buf.clear()
                
    def on_connect(self,status):
        self.print_t("on_connect", status)
        self.connected = status
        if status:
            self.clearbuf()

    def pull(self,size=-1):
        self.print_t("pull")
        
        try:
            while size==-1 or size>0:
                if size>0:
                    size-=1
                ch = self.buf.popleft()
                self.readbuf.append( ch )
                # peek ahead
                while ch in [0x00,0x01,0x02,0x03,0x04,0x05,ord("\r"),ord("\n")]:            
                    ch = self.buf.popleft()
                    if ch in  [0x00,0x01,0x02,0x03,0x04,0x05,ord("\r"),ord("\n")]:
                        self.readbuf.append( ch )
                    else:
                        self.buf.appendleft(ch)
                        return
                        
        except IndexError:
            pass

    def read(self,size=-1):
        self.print_t("read")
        
        self.pull(size)
        
        r = bytearray(self.readbuf)
        self.readbuf.clear()
        
        self.print_d("read", r )
        
        return r if len(r)>0 else None
    
    def write(self,buf):
        self.print_t("write")
        
        wb = self.ws.send(buf)
        return wb

    def flush(self):
        self.print_t("flush")
        
        pass

    # ws app plumbing
    
    def on_message(self, message):
        if self.debug_msg or self.trace:
            print("(incoming)","on_message", message.encode())    
        self.buf.extend(message.encode())
        
    def on_error(self, err):
        self.print_t("on_error",err)
        self.ws.keep_running = False

    def on_close(self):
        self.print_t("on_close")
        self.on_connect( ws, False )

    def on_open(self):
        self.print_t("on_open")
        
        def login(*args):
                    
            time.sleep(1)
            r = self.read()            
            self.print_d("open", r)
            
            self.write( self.passwd + "\n")            
            time.sleep(1)
            r = self.read()
            self.print_d("open", r)
            
            connected = False
            
            if r == b'\r\n':
                r = self.read()
                self.print_d("open", r)
                if r == b'WebREPL connected\r\n':
                    r = self.read(4)
                    self.print_d("open", r)
                    
                    connected = ( r == b'>>> ' )
                    
            self.on_connect( connected )
            
        # start logon task
        connectionThread = Thread( None, login, args=[] )
        connectionThread.start()
        return connectionThread


    def loop(self):
        
        def run_app(*args):
            r = self.ws.run_forever()
            self.print_d( "ws app finished with", r )
             
        loopThread = Thread( None, run_app, args=[] )
        loopThread.start()
        
        return loopThread


def pttywsappopen(url,passwd,timeout=3, debug=False, trace=False, debug_msg=False):
    
    wsapp = _WebSocketApp_Dev(passwd=passwd,debug=debug,trace=trace,debug_msg=debug_msg)
    wsapp.ws = websocket.WebSocketApp(url, 
                                on_message = wsapp.on_message,
                                on_error = wsapp.on_error,
                                on_close = wsapp.on_close,
                                on_open = wsapp.on_open,
                                )
    # start thread loop
    thrd = wsapp.loop()
    
    while wsapp.connected is None:
        if debug:
            print( "waiting for connection" )
        time.sleep( 0.5 )
    
    return wsapp


if __name__ == "__main__":
     
    url, passwd = "ws://192.168.178.21:8266","123456"
     
    with pttywsappopen( url, passwd, debug=True, trace=True, debug_msg=False ) as wsapp:    
        time.sleep(7)    
        print("request shutdown of all threads")
        
    

pttydev/sample.py

from pttydev import *

try:
    from dev_serial import *
    from dev_ws import *
    from dev_wsapp import *

except:
    pass

# open wrappers

# a open function is required to reconnect properly the device in case of an error

def get_pttyopen():

    port = '/dev/ttyUSB0' 
    baudrate = 115200
    bytesize = 8
    parity = 'N'
    stopbits = 1
    timeout = 0.25

    return pttyopen(port=port, baudrate=baudrate,
                    bytesize=bytesize, parity=parity, stopbits=stopbits,
                    timeout=timeout)

# put here the board IP adr
IP_ADR = "192.168.178.30"
IP_ADR = "192.168.178.26"

IP_PORT = 8266
IP_PASS = "123456"

def get_pttywsopen():
    # check this !!!
    # ESP32 delay 0.3-0.5
    # ESP8266 delay 0.01
    delay = 0.3
    return pttywsopen( f"ws://{IP_ADR}:{IP_PORT}", IP_PASS, write_delay=delay )

##
## dont use this one !!!
##
## deprecated, will be removed soon
##
def get_pttywsappopen():    
    return pttywsappopen( f"ws://{IP_ADR}:{IP_PORT}", IP_PASS, debug=False)
##
##
##

# the sample code

def send_cntrl_c(ptty,timeout=1):
    ptty.reset_input_buffer()
    ptty.write([0x03,0x03])
    buf = []
    while True:
        r = ptty.readline(timeout=timeout)
        if r is None: # timeout
            break
        buf.append(r)
        if r.startswith(b'>>>'):
            break
    return buf
        
def enter_raw_repl(ptty,timeout=1):
    ptty.write([0x01])
    r = ptty.readline(timeout=timeout)
    if r.startswith(b'raw REPL;'):
        r = ptty.read(1)
        if r == b'>':
            return True
    raise Exception("could not enter raw-repl")

def skipOK(ptty,timeout=1):
    print("skipOK")
    r = ptty.read(2,timeout=timeout)
    print("skipOK", r )
    if r!=b'OK':
        raise Exception("")
    return r

def skip0x04(ptty,timeout=1):
    print("skip0x04")
    buf = bytearray()
    while True:
        ch = ptty.read(1,timeout=timeout) # this performs slow!
        if ch == b'\x04':
            break
        buf.extend(ch)
    return buf

def send_cmd(ptty,cmd,handshake=True,echocmd=False,timeout=3):
    print("send_cmd")
    import textwrap
    cmd = textwrap.dedent( cmd )
    if echocmd:
        print(cmd)
    ptty.write(cmd.encode())
    ptty.write([0x04])
    if handshake:
        skipOK(ptty,timeout=timeout)
        data = skip0x04(ptty,timeout=timeout)
        err = skip0x04(ptty,timeout=timeout)
        r = ptty.read(1,timeout=timeout)
        if r != b'>':
            raise Exception("no prompt",data,err)
        return data,err


def sample():
    
    """this sample code opens the ptty, performs some commands and follows the output"""
    
    tty = PseudoTTY(
                    # uncomment the part you want to test
                    #get_pttyopen,
                    get_pttywsopen,
                    #get_pttywsappopen,
                    
                    #thrd_reader=_the_thread_reader,
                    
                    timeout=0.35,
                    block_size=512,
                    reconnect_delay=1,
                    
                    debug=True,
                    #trace_on=True,                    
                    #thrd_debug=True,
                    thrd_trace_on=True
                    )

    with tty.open() as ptty:
        
        ptty.waitready() # raises PseudoTTYOpenException 
        
        print("ptty", ptty )
        
        try:
            
            send_cntrl_c(ptty)
            enter_raw_repl(ptty)
            
            r = send_cmd(ptty,"""
                    print('hello world!')
                """)
            print( r )
 
            r = send_cmd(ptty,"""
                    import uos
                    print('12345')
                    print(uos.listdir())
                """)
            print( r )

            r = send_cmd(ptty,"""
                    import uos
                    print('12345')
                    print(uos.listdir())
                """,echocmd=True)
            print( r )

            r = send_cmd(ptty,"""
                    import uos
                    _m=False
                    _r=False
                    try:
                        uos.mkdir("test_dir")
                        _m=True
                    except:
                        pass
                    try:
                        uos.rmdir("test_dir")
                        _r=True
                    except:
                        pass
                    print(_m,_r)
                """)
            print( r )

            r = send_cmd(ptty,"""
                    import json, machine, sys, network, ubinascii
                    _ids = { 'dummy': True }
                    _ids["unique_id"] = ubinascii.hexlify( machine.unique_id() )
                    _ids["platform"] = sys.platform
                    _ids["version"] = sys.version
                    #########################
                    #########################
                    ## long cmd content
                    #########################
                    #########################
                    ###
                    print( json.dumps( _ids ) )
                """,echocmd=True)
            print( r )
            
            #return

            try:
                r = send_cmd(ptty, """
                        #
                        # long running task
                        #
                        for i in range(1,1000):
                            print(i)
                        #
                        """ , echocmd=True)                    
                print( r )
            except Exception as e:
                print("runtime error", ex )
                print(type(ex))
                print(ex.args)

            print("reset micropython")
            ptty.write([0x03,0x04]) # cntrl c and cntrl d 

            while True:                
                # get the current size of available chars in the internal buffer
                # if no crlf is in the buffer readline will not return content (or if timed out)
                available = len(ptty) 
                if available>0:
                    try:
                        r = ptty.readline(timeout=3)
                        print("main",r)
                    except PseudoTTYTimeoutException:
                        print("timeout")                       

        except KeyboardInterrupt as intr:
            print("user stop request", intr )        
        except Exception as ex:
            print("runtime error", ex )
            print(type(ex))
            print(ex.args)
            import traceback
            traceback.print_last()
        finally:
            print("close device")
        
    return
    

if __name__=='__main__':
    print( "running", __name__ )
    sample()
    
    